save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/service/LokiLogService.kt

```diff
@@ -0,0 +1,87 @@
+package com.saveourtool.save.service
+
+import com.saveourtool.save.utils.sortBy
+
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.fasterxml.jackson.databind.node.ObjectNode
+import org.springframework.web.reactive.function.client.WebClient
+import org.springframework.web.reactive.function.client.bodyToMono
+import reactor.core.publisher.Flux
+import reactor.kotlin.core.publisher.toFlux
+
+import java.time.Instant
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+import java.time.format.DateTimeFormatterBuilder
+import java.time.temporal.ChronoField
+
+/**
+ * AgentLogService from Loki
+ */
+class LokiLogService(
+    lokiServiceUrl: String,
+) : LogService {
+    private val webClient = WebClient.create(lokiServiceUrl)
+
+    override fun get(containerName: String, from: Instant, to: Instant): Flux<String> = webClient.get()
+        .uri(
+            "/loki/api/v1/query_range?query={query}&start={start}&end={end}&direction=forward",
+            "{container_name=\"$containerName\"}",
+            from.toEpochNanoStr(),
+            to.toEpochNanoStr(),
+        )
+        .retrieve()
```

Comment:
Does Loki do any batching of data? I remember that frontend (Grafana) sometimes complains if a single entry is too large, not sure about if there are too many values

Reply:
yes, it is. also it has limit and now it's set to 100 by default.
checking how to operate with batches

Reply:
checked in `lokicli`: there is no built-in batch mechanism, but they apply the following approach:
* query with limit=batch size using provided start_time and end_time
* take time from last entry from received result and set it to start_time
* repeat request till reach end_time or receive an empty result on new request

We have to implement similar (I think making limit=MAX_INT is not proper way)

Reply:
implemented similar logic as in loki/logcli